File: src/lib/merkle/tile-log.ts
/**
 * Tile-Based Merkle Tree Implementation
 * RFC 6962 compliant Merkle tree with C2SP tlog-tiles storage format
 */

import type { Storage } from "../storage/interface.ts";
import {
  TILE_SIZE,
  HASH_SIZE,
  tileIndexToPath,
  entryTileIndexToPath,
  entryIdToTileIndex,
  entryIdToTileOffset,
} from "./tile-naming.ts";

/**
 * Tile Log State
 * Stored in object storage for persistence
 */
interface TileLogState {
  size: number; // Number of leaves in the tree
  root: Uint8Array | null; // Current tree root hash
}

/**
 * Tile-based Merkle tree
 * Implements RFC 6962 Merkle tree with C2SP tlog-tiles storage
 */
export class TileLog {
  private storage: Storage;
  private state: TileLogState;

  constructor(storage: Storage) {
    this.storage = storage;
    this.state = {
      size: 0,
      root: null,
    };
  }

  /**
   * Load tree state from storage
   * Called on initialization to restore tree from persistent storage
   */
  async load(): Promise<void> {
    try {
      const stateData = await this.storage.get(".tree-state");
      if (stateData) {
        const decoded = JSON.parse(new TextDecoder().decode(stateData));
        this.state.size = decoded.size;
        this.state.root = decoded.root ? new Uint8Array(decoded.root) : null;
      }
    } catch (error) {
      // Empty tree - no state file exists yet
      this.state = { size: 0, root: null };
    }
  }

  /**
   * Save tree state to storage
   */
  private async saveState(): Promise<void> {
    const encoded = JSON.stringify({
      size: this.state.size,
      root: this.state.root ? Array.from(this.state.root) : null,
    });
    await this.storage.put(".tree-state", new TextEncoder().encode(encoded));
  }

  /**
   * Append a leaf to the tree
   * Returns the entry ID of the appended leaf
   */
  async append(leaf: Uint8Array): Promise<number> {
    if (leaf.length !== HASH_SIZE) {
      throw new Error(`Invalid leaf size: expected ${HASH_SIZE} bytes, got ${leaf.length}`);
    }

    const entryId = this.state.size;

    // Store in entry tile
    await this.appendToEntryTile(entryId, leaf);

    // Update hash tiles
    await this.updateHashTiles(entryId, leaf);

    // Increment size and recompute root
    this.state.size++;
    this.state.root = await this.computeRoot();

    // Persist state
    await this.saveState();

    return entryId;
  }

  /**
   * Get the current tree size (number of leaves)
   */
  async size(): Promise<number> {
    return this.state.size;
  }

  /**
   * Get the current tree root hash
   */
  async root(): Promise<Uint8Array> {
    if (this.state.root === null) {
      throw new Error("Cannot get root of empty tree");
    }
    return this.state.root;
  }

  /**
   * Get a leaf by entry ID
   */
  async getLeaf(entryId: number): Promise<Uint8Array> {
    if (entryId >= this.state.size) {
      throw new Error(`Entry ID ${entryId} out of bounds (size: ${this.state.size})`);
    }

    const tileIndex = entryIdToTileIndex(entryId);
    const tileOffset = entryIdToTileOffset(entryId);

    const tilePath = entryTileIndexToPath(tileIndex);
    const tileData = await this.storage.get(tilePath);

    if (!tileData) {
      throw new Error(`Entry tile not found: ${tilePath}`);
    }

    // Extract the specific hash from the tile
    const start = tileOffset * HASH_SIZE;
    const end = start + HASH_SIZE;

    return tileData.slice(start, end);
  }

  /**
   * Append leaf to entry tile
   * Entry tiles store the raw leaf data
   */
  private async appendToEntryTile(entryId: number, leaf: Uint8Array): Promise<void> {
    const tileIndex = entryIdToTileIndex(entryId);
    const tilePath = entryTileIndexToPath(tileIndex);

    // Read existing tile (if any)
    const existingTile = await this.storage.get(tilePath);
    const currentSize = existingTile ? existingTile.length / HASH_SIZE : 0;

    // Append new leaf
    const newTile = new Uint8Array((currentSize + 1) * HASH_SIZE);
    if (existingTile) {
      newTile.set(existingTile, 0);
    }
    newTile.set(leaf, currentSize * HASH_SIZE);

    // Write updated tile
    await this.storage.put(tilePath, newTile);
  }

  /**
   * Update hash tiles after appending a leaf
   * Hash tiles store the Merkle tree structure
   */
  private async updateHashTiles(entryId: number, leaf: Uint8Array): Promise<void> {
    // Level 0: Store leaf hash in tile
    await this.appendToHashTile(0, entryId, leaf);

    // Propagate up the tree if we completed a tile
    const tileOffset = entryIdToTileOffset(entryId);
    if (tileOffset === TILE_SIZE - 1) {
      // We just completed a tile - compute its hash and propagate
      await this.propagateTileHash(0, entryIdToTileIndex(entryId));
    }
  }

  /**
   * Append hash to a tile at a specific level
   */
  private async appendToHashTile(level: number, entryId: number, hash: Uint8Array): Promise<void> {
    const tileIndex = entryIdToTileIndex(entryId);
    const tilePath = tileIndexToPath(level, tileIndex);

    // Read existing tile (if any)
    const existingTile = await this.storage.get(tilePath);
    const currentSize = existingTile ? existingTile.length / HASH_SIZE : 0;

    // Append new hash
    const newTile = new Uint8Array((currentSize + 1) * HASH_SIZE);
    if (existingTile) {
      newTile.set(existingTile, 0);
    }
    newTile.set(hash, currentSize * HASH_SIZE);

    // Write updated tile
    await this.storage.put(tilePath, newTile);
  }

  /**
   * Propagate completed tile hash to the next level
   */
  private async propagateTileHash(level: number, tileIndex: number): Promise<void> {
    // Read the completed tile
    const tilePath = tileIndexToPath(level, tileIndex);
    const tileData = await this.storage.get(tilePath);

    if (!tileData) {
      throw new Error(`Tile not found: ${tilePath}`);
    }

    // Compute hash of the tile
    const tileHash = await this.hashData(tileData);

    // Store in next level
    const nextLevel = level + 1;
    const nextEntryId = tileIndex; // Tile index becomes entry ID at next level

    await this.appendToHashTile(nextLevel, nextEntryId, tileHash);

    // Recursively propagate if we completed a tile at the next level
    const nextTileOffset = entryIdToTileOffset(nextEntryId);
    if (nextTileOffset === TILE_SIZE - 1) {
      await this.propagateTileHash(nextLevel, entryIdToTileIndex(nextEntryId));
    }
  }

  /**
   * Compute the current tree root
   * RFC 6962 MTH (Merkle Tree Hash) computation
   */
  private async computeRoot(): Promise<Uint8Array> {
    if (this.state.size === 0) {
      throw new Error("Cannot compute root of empty tree");
    }

    if (this.state.size === 1) {
      // Root is just the single leaf (RFC 6962: MTH({d[0]}) = SHA-256(0x00 || d[0]))
      const leaf = await this.getLeaf(0);
      return await this.hashLeaf(leaf);
    }

    // For larger trees, compute from tiles
    return await this.computeRootFromTiles();
  }

  /**
   * Compute root from tile structure
   * Combines full and partial tiles at each level
   */
  private async computeRootFromTiles(): Promise<Uint8Array> {
    let currentLevel = 0;
    let currentSize = this.state.size;

    // Find the highest level with data
    while (currentSize > 1) {
      const numTiles = Math.ceil(currentSize / TILE_SIZE);

      if (numTiles === 1) {
        // Single partial tile at this level - hash it to get root
        const tilePath = tileIndexToPath(currentLevel, 0);
        const tileData = await this.storage.get(tilePath);

        if (!tileData) {
          throw new Error(`Tile not found: ${tilePath}`);
        }

        // Hash all entries in the tile to get the root
        return await this.hashTileToRoot(tileData, currentSize);
      }

      // Move up one level
      currentLevel++;
      currentSize = numTiles;
    }

    // Single tile at current level
    const tilePath = tileIndexToPath(currentLevel, 0);
    const tileData = await this.storage.get(tilePath);

    if (!tileData) {
      throw new Error(`Root tile not found: ${tilePath}`);
    }

    return tileData.slice(0, HASH_SIZE);
  }

  /**
   * Hash a tile to compute root
   * Combines hashes using RFC 6962 algorithm
   * Note: Tile contains raw leaf data, so we need to hash each leaf first
   */
  private async hashTileToRoot(tileData: Uint8Array, numHashes: number): Promise<Uint8Array> {
    if (numHashes === 0) {
      throw new Error("Cannot hash empty tile");
    }

    if (numHashes === 1) {
      // Single leaf - hash it with 0x00 prefix
      const leaf = tileData.slice(0, HASH_SIZE);
      return await this.hashLeaf(leaf);
    }

    // Build a binary tree from the leaves in the tile
    // First, hash each leaf with 0x00 prefix
    const hashes: Uint8Array[] = [];
    for (let i = 0; i < numHashes; i++) {
      const start = i * HASH_SIZE;
      const leaf = tileData.slice(start, start + HASH_SIZE);
      const leafHash = await this.hashLeaf(leaf);
      hashes.push(leafHash);
    }

    // Recursively combine hashes
    return await this.mergeHashes(hashes);
  }

  /**
   * Recursively merge hashes into a single root
   * RFC 6962: MTH(D[n]) = SHA-256(0x01 || MTH(D[0:k]) || MTH(D[k:n]))
   * where k is the largest power of 2 less than n
   */
  private async mergeHashes(hashes: Uint8Array[]): Promise<Uint8Array> {
    const n = hashes.length;

    if (n === 0) {
      throw new Error("Cannot merge empty hash list");
    }

    if (n === 1) {
      return hashes[0];
    }

    if (n === 2) {
      // Base case: combine two hashes
      return await this.hashNode(hashes[0], hashes[1]);
    }

    // Find k: largest power of 2 less than n
    const k = this.largestPowerOfTwoLessThan(n);

    const left = hashes.slice(0, k);
    const right = hashes.slice(k);

    const leftHash = await this.mergeHashes(left);
    const rightHash = await this.mergeHashes(right);

    return await this.hashNode(leftHash, rightHash);
  }

  /**
   * Find largest power of 2 strictly less than n
   * For RFC 6962 MTH algorithm
   */
  private largestPowerOfTwoLessThan(n: number): number {
    let k = 1;
    while (k * 2 < n) {
      k *= 2;
    }
    return k;
  }

  /**
   * Hash a leaf with RFC 6962 prefix (0x00)
   */
  private async hashLeaf(leaf: Uint8Array): Promise<Uint8Array> {
    const data = new Uint8Array(1 + leaf.length);
    data[0] = 0x00; // RFC 6962 leaf prefix
    data.set(leaf, 1);
    return await this.hashData(data);
  }

  /**
   * Hash an internal node with RFC 6962 prefix (0x01)
   */
  private async hashNode(left: Uint8Array, right: Uint8Array): Promise<Uint8Array> {
    const data = new Uint8Array(1 + left.length + right.length);
    data[0] = 0x01; // RFC 6962 node prefix
    data.set(left, 1);
    data.set(right, 1 + left.length);
    return await this.hashData(data);
  }

  /**
   * Hash data using SHA-256
   */
  private async hashData(data: Uint8Array): Promise<Uint8Array> {
    const hashBuffer = await crypto.subtle.digest("SHA-256", data);
    return new Uint8Array(hashBuffer);
  }
}

File: scitt-golang/pkg/merkle/tilelog.go
package merkle

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"

	"github.com/tradeverifyd/transparency-service/scitt-golang/pkg/storage"
	"golang.org/x/mod/sumdb/tlog"
)

// TileLog represents a tile-based Merkle tree
// Implements RFC 6962 Merkle tree with C2SP tlog-tiles storage
type TileLog struct {
	storage storage.Storage
	size    int64
}

// TileLogState represents the persistent state of the tree
type TileLogState struct {
	Size int64  `json:"size"`
	Root []byte `json:"root,omitempty"`
}

// NewTileLog creates a new tile log with the given storage
func NewTileLog(storage storage.Storage) *TileLog {
	return &TileLog{
		storage: storage,
		size:    0,
	}
}

// Load loads the tree state from storage
func (tl *TileLog) Load() error {
	stateData, err := tl.storage.Get(".tree-state")
	if err != nil {
		return fmt.Errorf("failed to get tree state: %w", err)
	}

	if stateData == nil {
		// Empty tree - no state file exists yet
		tl.size = 0
		return nil
	}

	var state TileLogState
	if err := json.Unmarshal(stateData, &state); err != nil {
		return fmt.Errorf("failed to unmarshal tree state: %w", err)
	}

	tl.size = state.Size
	return nil
}

// saveState saves the tree state to storage
func (tl *TileLog) saveState() error {
	var root []byte
	if tl.size > 0 {
		rootHash, err := tl.TreeHash(tl.size)
		if err != nil {
			return fmt.Errorf("failed to compute tree hash: %w", err)
		}
		root = rootHash[:]
	}

	state := TileLogState{
		Size: tl.size,
		Root: root,
	}

	stateData, err := json.Marshal(state)
	if err != nil {
		return fmt.Errorf("failed to marshal tree state: %w", err)
	}

	if err := tl.storage.Put(".tree-state", stateData); err != nil {
		return fmt.Errorf("failed to put tree state: %w", err)
	}

	return nil
}

// Append appends a leaf to the tree
// Returns the entry ID of the appended leaf
// The leaf should be the raw record hash (e.g., SHA-256 of statement)
// This function will apply the RFC 6962 leaf prefix (0x00) before storing
func (tl *TileLog) Append(leaf [HashSize]byte) (int64, error) {
	entryID := tl.size

	// Apply RFC 6962 leaf hash prefix (0x00) to the raw record hash
	// This converts the raw hash into a proper Merkle tree leaf hash
	leafHash := hashLeaf(leaf)

	// Store the leaf hash (with RFC 6962 prefix applied) in entry tile
	if err := tl.appendToEntryTile(entryID, leafHash); err != nil {
		return 0, fmt.Errorf("failed to append to entry tile: %w", err)
	}

	// Increment size AFTER appending so TreeHash can read the new leaf
	// TreeHash(n) expects to be able to read leaves 0 through n-1
	// We increment here so the new leaf at entryID is included in tl.size
	tl.size++

	// Persist state (this will compute TreeHash(tl.size) which includes the new leaf)
	if err := tl.saveState(); err != nil {
		// Roll back size increment on failure
		tl.size--
		return 0, fmt.Errorf("failed to save state: %w", err)
	}

	return entryID, nil
}

// Size returns the current tree size (number of leaves)
func (tl *TileLog) Size() int64 {
	return tl.size
}

// Root returns the current tree root hash
func (tl *TileLog) Root() ([HashSize]byte, error) {
	if tl.size == 0 {
		return [HashSize]byte{}, fmt.Errorf("cannot get root of empty tree")
	}

	return tl.TreeHash(tl.size)
}

// GetLeaf retrieves a leaf by entry ID
func (tl *TileLog) GetLeaf(entryID int64) ([HashSize]byte, error) {
	if entryID >= tl.size {
		return [HashSize]byte{}, fmt.Errorf("entry ID %d out of bounds (size: %d)", entryID, tl.size)
	}

	return tl.getLeafDirect(entryID)
}

// getLeafDirect retrieves a leaf by entry ID without bounds checking
func (tl *TileLog) getLeafDirect(entryID int64) ([HashSize]byte, error) {
	tileIndex := EntryIDToTileIndex(entryID)
	tileOffset := EntryIDToTileOffset(entryID)

	tilePath := EntryTileIndexToPath(tileIndex, nil)
	tileData, err := tl.storage.Get(tilePath)
	if err != nil {
		return [HashSize]byte{}, fmt.Errorf("failed to get entry tile: %w", err)
	}

	if tileData == nil {
		return [HashSize]byte{}, fmt.Errorf("entry tile not found: %s", tilePath)
	}

	// Extract the specific hash from the tile
	start := tileOffset * HashSize
	end := start + HashSize

	if end > len(tileData) {
		return [HashSize]byte{}, fmt.Errorf("tile data too short for offset %d", tileOffset)
	}

	var leaf [HashSize]byte
	copy(leaf[:], tileData[start:end])
	return leaf, nil
}

// appendToEntryTile appends a leaf to an entry tile
func (tl *TileLog) appendToEntryTile(entryID int64, leaf [HashSize]byte) error {
	tileIndex := EntryIDToTileIndex(entryID)
	tilePath := EntryTileIndexToPath(tileIndex, nil)

	// Read existing tile (if any)
	existingTile, err := tl.storage.Get(tilePath)
	if err != nil {
		return fmt.Errorf("failed to get existing tile: %w", err)
	}

	var currentSize int
	if existingTile != nil {
		currentSize = len(existingTile) / HashSize
	}

	// Append new leaf
	newTile := make([]byte, (currentSize+1)*HashSize)
	if existingTile != nil {
		copy(newTile, existingTile)
	}
	copy(newTile[currentSize*HashSize:], leaf[:])

	// Write updated tile
	if err := tl.storage.Put(tilePath, newTile); err != nil {
		return fmt.Errorf("failed to put tile: %w", err)
	}

	return nil
}

// ReadHashes implements tlog.HashReader interface
// Reads hashes from tree at given indexes
// Note: tlog uses 0-indexed record IDs (0 through n-1 for n records)
// but C2SP tlog-tiles uses 1-indexed storage (records 1 through n)
func (tl *TileLog) ReadHashes(indexes []int64) ([]tlog.Hash, error) {
	hashes := make([]tlog.Hash, len(indexes))

	for i, index := range indexes {
		// Convert from tlog's 0-indexed to our 0-indexed storage
		// tlog requests record at index i, we store it at entry ID i
		leaf, err := tl.getLeafDirect(index)
		if err != nil {
			return nil, fmt.Errorf("failed to get leaf %d: %w", index, err)
		}

		// Convert to tlog.Hash
		var hash tlog.Hash
		copy(hash[:], leaf[:])
		hashes[i] = hash
	}

	return hashes, nil
}

// TreeHash computes the tree hash for a given tree size using tlog
func (tl *TileLog) TreeHash(n int64) (tlog.Hash, error) {
	if n == 0 {
		return tlog.Hash{}, fmt.Errorf("cannot compute hash of empty tree")
	}

	if n > tl.size {
		return tlog.Hash{}, fmt.Errorf("requested size %d exceeds tree size %d", n, tl.size)
	}

	// tlog.TreeHash expects records numbered 0..n-1
	// It will call ReadHashes with those indices
	// We provide those via our getLeafDirect which maps directly to entry IDs
	return tlog.TreeHash(n, tl)
}

// RecordHash computes the hash of a record (leaf) with RFC 6962 prefix
func RecordHash(data []byte) [HashSize]byte {
	h := sha256.New()
	h.Write([]byte{0x00}) // RFC 6962 leaf prefix
	h.Write(data)
	var result [HashSize]byte
	copy(result[:], h.Sum(nil))
	return result
}

// tileHashStorage implements tlog.TileReader for hash tiles
type tileHashStorage struct {
	storage storage.Storage
}

// Height returns the tile height (always 8 for standard tiles)
func (ths *tileHashStorage) Height() int {
	return 8 // 2^8 = 256 hashes per tile
}

// ReadTiles reads tiles from storage
func (ths *tileHashStorage) ReadTiles(tiles []tlog.Tile) ([][]byte, error) {
	data := make([][]byte, len(tiles))

	for i, tile := range tiles {
		// Determine if this is a partial tile
		var path string
		if tile.W != 1<<uint(ths.Height()) {
			// Partial tile
			width := int(tile.W)
			path = TileIndexToPath(tile.L, tile.N, &width)
		} else {
			// Full tile
			path = TileIndexToPath(tile.L, tile.N, nil)
		}

		tileData, err := ths.storage.Get(path)
		if err != nil {
			return nil, fmt.Errorf("failed to get tile %s: %w", path, err)
		}

		data[i] = tileData
	}

	return data, nil
}

// SaveTiles saves tiles to storage
func (ths *tileHashStorage) SaveTiles(tiles []tlog.Tile, data [][]byte) {
	for i, tile := range tiles {
		// Determine if this is a partial tile
		var path string
		if tile.W != 1<<uint(ths.Height()) {
			// Partial tile
			width := int(tile.W)
			path = TileIndexToPath(tile.L, tile.N, &width)
		} else {
			// Full tile
			path = TileIndexToPath(tile.L, tile.N, nil)
		}

		// Ignore errors in SaveTiles (as per tlog.TileReader interface)
		_ = ths.storage.Put(path, data[i])
	}
}
